File: java/src/leetcode_linked_lists/easy/easy/reverse_linked_list.java
/**
 * 
 * 
 * Given the head of a singly linked list, reverse the list, and return the reversed list.

 
 */

 /**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head;
        ListNode prev = null;
        while(curr != null){
            ListNode temp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = temp;
        }
        
        return prev;
    }
    private ListNode recursiveReverse(ListNode head){
        if (head == null || head.next == null) return head;
        ListNode p = recursiveReverse(head.next);
        head.next.next = head;
        head.next = null;
        return p;
    }
}

File: python/Data Structures/Linked Lists/reverseLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # if head.next == None:
        #     return head
        
        prev = None
        curr = head
        
        while curr != None:
            # tmp = curr.next
            # curr.next = prev
            # tmp.next = curr
            # curr = tmp
            
            temp = head
            print("temp", temp)
            head.next = curr.next
            curr.next = head
            # curr = curr.next
            
            print("HeadAfterNext", head)
            print("CurrAfterNext", curr)
            
            head = curr
            print("head", head)
            curr = temp.next
            print("curr", curr)
            
            curr2 = head
            while curr2 != None:
                print(curr2.val)
                curr2 = curr2.next
            print("-------------")
            
            
            
            # nextNode = curr.next
            
            
        return head
        